### PR TITLE
DEP: linalg: deprecate kron

### LINKED ISSUE #20072: ENH:linalg: kron does not support nd arrays with n>2
Is this a feature?

> Re. `scipy.linalg.kron`, I'm not sure whether it might not be better to just delegate to `numpy.kron`, unless there are any advantages to `scipy.linalg.kron`. `numpy.kron` works with empty array arguments and also handles n-dimensional arrays, whereas `scipy.linalg.kron` gives some funny results when the number of dimensions is not two:
>
>```python
>linalg.kron(np.empty((1,2,3)), >np.empty((4,5,6))).shape
>(3, 8, 5, 6)  # should be (4, 10, 18)
>```

_Originally posted by @JozsefKutas in https://github.com/scipy/scipy/pull/18673#issuecomment-1590329028_

### LINKED ISSUE #20077: DEP: linalg: deprecate `kron`
`linalg.kron` is a basic, data-movement intensive array operation. There is hardly any computation done other than scalar times array and quite some shuffling needed for higher order nd arrays. It is similar to a very special case of `np.einsum` in higher dimensions. Hence it doesn't make too much sense for us to carry around a suboptimal version of it since Numpy already supports batch array input https://numpy.org/doc/stable/reference/generated/numpy.kron.html

For comparison we also have Khatri-Rao product which is a Kronecker-like product and it is specific enough for us to have it but kron is not in my opinion.

See #20072 and https://github.com/pymc-devs/pytensor/issues/640 for more context.

### PR DESCRIPTION
**Pull request by j-bowhay**
#### Reference issue

closes #20077
closes #20072
#### What does this implement/fix?

This was previously blocked by array-api discussions however now with https://github.com/data-apis/array-api-extra/pull/7 I think we can now proceed.

### CODE REVIEW

File: scipy/linalg/_special_matrices.py [hunk 1]
```diff
@@ -339,6 +344,9 @@ def kron(a, b):
            [3, 3, 3, 4, 4, 4]])
 
     """
+    msg = ("`kron` has been deprecated infavour of `numpy.kron` in SciPy"
```
Comment (lucascolley) [inline scipy/linalg/_special_matrices.py:347]:
```suggestion
    msg = ("`kron` has been deprecated in favour of `numpy.kron` in SciPy"
```

**Review by lucascolley — APPROVED** [review]
thanks! just one comment

**Comment by lucascolley** [comment]
doctest failure is real

**Review by tylerjereddy — COMMENTED** [review]
For the doctest failure, this seems to do the trick locally (a bit silly, though I suppose it makes the deprecation even harder to miss...). Maybe Evgeni has a slicker way to suppress the failure.

```diff
--- a/scipy/linalg/_special_matrices.py
+++ b/scipy/linalg/_special_matrices.py
@@ -337,9 +337,12 @@ def kron(a, b):
 
     Examples
     --------
+    >>> import warnings
     >>> from numpy import array
     >>> from scipy.linalg import kron
-    >>> kron(array([[1,2],[3,4]]), array([[1,1,1]]))
+    >>> with warnings.catch_warnings():
+    ...     warnings.simplefilter("ignore", DeprecationWarning)
+    ...     kron(array([[1,2],[3,4]]), array([[1,1,1]]))
     array([[1, 1, 1, 2, 2, 2],
            [3, 3, 3, 4, 4, 4]])
```

**Comment by ev-br** [comment]
There's a DeprecationWarning filter for doctests in https://github.com/scipy/scipy/blob/main/scipy/conftest.py#L382-L389, if you want to use that.

**Comment by lucascolley** [comment]
thanks Jake!

**Pull request merged**